buildpacks/nodejs-npm-engine/src/main.rs

```diff
@@ -0,0 +1,102 @@
+mod cmd;
+mod errors;
+mod layers;
+
+use crate::errors::NpmEngineBuildpackError;
+use crate::layers::npm_engine::NpmEngineLayer;
+use heroku_nodejs_utils::inv::Inventory;
+use heroku_nodejs_utils::package_json::PackageJson;
+use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+use libcnb::data::build_plan::BuildPlanBuilder;
+use libcnb::data::layer_name;
+use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use libcnb::generic::{GenericMetadata, GenericPlatform};
+use libcnb::{buildpack_main, Buildpack, Env};
+use libherokubuildpack::log::{log_header, log_info};
+
+const INVENTORY: &str = include_str!("../inventory.toml");
+
+pub(crate) struct NpmEngineBuildpack;
+
+impl Buildpack for NpmEngineBuildpack {
+    type Platform = GenericPlatform;
+    type Metadata = GenericMetadata;
+    type Error = NpmEngineBuildpackError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        let package_json = context.app_dir.join("package.json");
+        package_json
+            .exists()
+            .then(|| {
+                PackageJson::read(package_json)
+                    .map(|package_json| package_json.engines)
+                    .unwrap_or(None)
+                    .and_then(|engines| engines.npm)
+            })
+            .flatten()
+            .map(|_| {
+                DetectResultBuilder::pass()
+                    .build_plan(
+                        BuildPlanBuilder::new()
+                            .requires("npm")
+                            .requires("node")
+                            .provides("npm")
+                            .build(),
+                    )
+                    .build()
+            })
+            .unwrap_or_else(|| DetectResultBuilder::fail().build())
+    }
+
+    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+        log_header("Heroku npm Engine Buildpack");
+
+        let env = Env::from_current();
+
+        let package_json = PackageJson::read(context.app_dir.join("package.json"))
+            .map_err(NpmEngineBuildpackError::PackageJson)?;
+
+        let npm_requirement = package_json
```

Comment:
I think it's probably good info to output the version requirement. Looks like the logs immediately jump to `Resolved npm version: 9.8.0`. It think it's useful to print the semver range we're using too, something like:

```
Detected npm version requirement: 9.x
Resolved npm version: 9.8.1
```

The classic buildpack also annotated where the requirement was found, which might be nice too: `Detected npm version requirement: 9.x (from package.json engines)`

Reply:
👍 i went light on the output UX for these initial drafts (just enough to verify with tests). i'll incorporate this feedback into upcoming logging and error polish.

Reply:
@joshwlewis see the example output attached to the description